File: builder/Builder.ts

class Laptop {
    private cpu: string = "";
    private ram: string = "";
    private os: string = "";

    setCpu(cpu: string) {
        this.cpu = cpu;
    }

    setRam(ram: string) {
        this.ram = ram;
    }

    setOs(os: string) {
        this.os = os;
    }

    describe() {
        console.log("CPU: " + this.cpu);
        console.log("RAM: " + this.ram);
        console.log("OS: " + this.os);
    }
}

abstract class LaptopBuilder {
    public abstract setCpu(): void;
    public abstract setRam(): void;
    public abstract setOs(): void;
    public abstract build(): Laptop;
}

class HighPerformanceLaptopBuilder extends LaptopBuilder {
    private highPerformanceLaptop: Laptop = new Laptop();

    public setCpu() {
        this.highPerformanceLaptop.setCpu("i7 CPU");
    }

    public setRam() {
        this.highPerformanceLaptop.setRam("16GB RAM");
    }

    public setOs() {
        this.highPerformanceLaptop.setOs("Windows 10");
    }

    public build(): Laptop {
        return this.highPerformanceLaptop;
    }
}

class LowPerformanceLaptopBuilder extends LaptopBuilder {
    private lowPerformanceLaptop: Laptop = new Laptop();

    public setCpu() {
        this.lowPerformanceLaptop.setCpu("i3 CPU");
    }

    public setRam() {
        this.lowPerformanceLaptop.setRam("4GB RAM");
    }

    public setOs() {
        this.lowPerformanceLaptop.setOs("Windows 7");
    }

    public build(): Laptop {
        return this.lowPerformanceLaptop;
    }
}

class Director {
    private laptopBuilder: LaptopBuilder;

    public setLaptopBuilder(laptopBuilder: LaptopBuilder) {
        this.laptopBuilder = laptopBuilder;
    }

    public getLaptop(): Laptop {
        this.laptopBuilder.setCpu();
        this.laptopBuilder.setRam();
        this.laptopBuilder.setOs();
        return this.laptopBuilder.build();
    }
}

const testLaptopBuilder = () => {
    const director = new Director();
    const highPerformanceLaptopBuilder = new HighPerformanceLaptopBuilder();
    director.setLaptopBuilder(highPerformanceLaptopBuilder);
    const highPerformanceLaptop = director.getLaptop();
    console.log("High performance laptop:");
    highPerformanceLaptop.describe();

    const lowPerformanceLaptopBuilder = new LowPerformanceLaptopBuilder();
    director.setLaptopBuilder(lowPerformanceLaptopBuilder);
    const lowPerformanceLaptop = director.getLaptop();
    console.log("Low performance laptop:");
    lowPerformanceLaptop.describe();
}

testLaptopBuilder();

File: builder/Builder.java
public class Builder {
    private class Laptop {
        String cpu;
        String ram;
        String os;

        public Laptop() {
            this.cpu = "";
            this.ram = "";
            this.os = "";
        }

        public void setCpu(String cpu) {
            this.cpu = cpu;
        }

        public void setRam(String ram) {
            this.ram = ram;
        }

        public void setOs(String os) {
            this.os = os;
        }

        public void describe() {
            System.out.println("CPU: " + cpu);
            System.out.println("RAM: " + ram);
            System.out.println("OS: " + os);
        }
    }

    private abstract class LaptopBuilder {
        public abstract void setCpu();

        public abstract void setRam();

        public abstract void setOs();

        public abstract Laptop build();
    }

    private class HighPerformanceLaptopBuilder extends LaptopBuilder {
        private Laptop highPerformanceLaptop;

        public HighPerformanceLaptopBuilder() {
            highPerformanceLaptop = new Laptop();
        }

        @Override
        public void setCpu() {
            highPerformanceLaptop.setCpu("i7 CPU");
        }

        @Override
        public void setRam() {
            highPerformanceLaptop.setRam("16GB RAM");
        }

        @Override
        public void setOs() {
            highPerformanceLaptop.setOs("Windows 10");
        }

        @Override
        public Laptop build() {
            return highPerformanceLaptop;
        }
    }

    private class LowPerformanceLaptopBuilder extends LaptopBuilder {
        private Laptop lowPerformanceLaptop;

        public LowPerformanceLaptopBuilder() {
            lowPerformanceLaptop = new Laptop();
        }

        @Override
        public void setCpu() {
            lowPerformanceLaptop.setCpu("i3 CPU");
        }

        @Override
        public void setRam() {
            lowPerformanceLaptop.setRam("4GB RAM");
        }

        @Override
        public void setOs() {
            lowPerformanceLaptop.setOs("Windows 7");
        }

        @Override
        public Laptop build() {
            return lowPerformanceLaptop;
        }
    }

    private class Director {
        private LaptopBuilder laptopBuilder = null;

        public void setLaptopBuilder(LaptopBuilder laptopBuilder) {
            this.laptopBuilder = laptopBuilder;
        }

        public Laptop getLaptop() {
            laptopBuilder.setCpu();
            laptopBuilder.setRam();
            laptopBuilder.setOs();
            return laptopBuilder.build();
        }
    }

    public void testLaptopBuilder() {
        Director director = new Director();
        LaptopBuilder highPerformanceLaptopBuilder = new HighPerformanceLaptopBuilder();
        director.setLaptopBuilder(highPerformanceLaptopBuilder);
        Laptop highPerformanceLaptop = director.getLaptop();
        System.out.println("High performance laptop:");
        highPerformanceLaptop.describe();

        LaptopBuilder lowPerformanceLaptopBuilder = new LowPerformanceLaptopBuilder();
        director.setLaptopBuilder(lowPerformanceLaptopBuilder);
        Laptop lowPerformanceLaptop = director.getLaptop();
        System.out.println("Low performance laptop:");
        lowPerformanceLaptop.describe();
    }

    public static void main(String[] args) {
        Builder builder = new Builder();
        builder.testLaptopBuilder();
    }
}